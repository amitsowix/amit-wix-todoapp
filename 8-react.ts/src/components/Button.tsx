import {FC} from 'react';
import {createUseStyles} from 'react-jss'
import {Classes} from 'jss';
import addIcon from "../assets/add.png";

interface ButtonProps {
    onClick: () => void;
}

const Button: FC<ButtonProps> = (props) : JSX.Element => {
    const classes = useStyles();
    return (
        <div className={classes.newItemButton} onClick={props.onClick}>
            <img alt="" src={addIcon} className={classes.addIcon}/>
            <div>New Item</div>
        </div>
    )            
}

const useStyles: () => Classes = createUseStyles({
    newItemButton: {
        alignItems: "center",
        background: 'orangered',
        display: "flex",
        width: "108px",
        padding: "15px",
        justifyContent: "space-between",
        borderRadius: "5px",
        color: "white",
        fontSize: "17px",
        cursor: "pointer"
    }, 
    
    addIcon: {
        height: "30px",
        filter: "brightness(0) invert(1)"
    }, 
});

export default Button