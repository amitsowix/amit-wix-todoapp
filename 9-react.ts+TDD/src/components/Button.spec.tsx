import React from 'react';
import renderer, {ReactTestRenderer, ReactTestRendererJSON} from 'react-test-renderer';
import {render, screen} from '@testing-library/react';
import Button from './Button';

const driver = {
    when: {
        mount: (callback: () => void): ReactTestRenderer => {
            const component: ReactTestRenderer = renderer.create(<Button onClick={callback}/>);
            return component;
        },
        render: (callback: () => void): void => {
            render(<Button onClick={callback}/>);
        }
    },
    get: {
      title: () => screen.getByText('New Item')
    }
}

describe('Button Render', () => {

    it('should render correctly', () => {

        const component = driver.when.mount(()=>{});

        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should be a div', () => {

        const component = driver.when.mount(()=>{});

        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        if (tree && !Array.isArray(tree)){
            expect(tree.type).toMatch('div');
        }
    })

    it('should render the text', () => {

        driver.when.render(()=>{});

        expect(driver.get.title()).toBeInTheDocument();
    })
})


describe('Button functionality test', () => {
    it('should call the callback on click', () => {

        const mockCallBack = jest.fn();
        const component: ReactTestRenderer = driver.when.mount(mockCallBack);

        let tree: ReactTestRendererJSON | ReactTestRendererJSON[] | null = component.toJSON();
        if (tree && !Array.isArray(tree)){
            tree.props.onClick();
            expect(mockCallBack.mock.calls.length).toEqual(1);
        }
    });
})
