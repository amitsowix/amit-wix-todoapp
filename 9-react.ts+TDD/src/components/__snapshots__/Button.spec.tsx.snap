// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render should render correctly 1`] = `
<div
  className="newItemButton-0-2-1"
  data-hook="new-item-button"
  onClick={[Function]}
>
  <img
    alt=""
    className="addIcon-0-2-2"
    src="test-file-stub"
  />
  <div>
    New Item
  </div>
</div>
`;
